### PR TITLE
Rework Filter

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

File: source/jmap/api.d [hunk 1]
```diff
@@ -9,10 +9,10 @@ version (SIL) {
         handlers.openModule("jmap");
         scope (exit) handlers.closeModule();
 
-        static foreach (T; AliasSeq!(Credentials, JmapSessionParams, Session, Mailbox, MailboxRights, MailboxSortProperty, Filter, FilterOperator, FilterOperatorKind, FilterCondition, Comparator, Account, AccountParams, AccountCapabilities, SessionCoreCapabilities, Contact, ContactGroup, ContactInformation, JmapFile, EmailAddress, Envelope, ContactAddress, ResultReference, JmapResponseError, EmailProperty, EmailBodyProperty))
+        static foreach (T; AliasSeq!(Credentials, JmapSessionParams, Session, Mailbox, MailboxRights, MailboxSortProperty, FilterHolder, FilterOperator, FilterOperatorKind, FilterCondition, Comparator, Account, AccountParams, AccountCapabilities, SessionCoreCapabilities, Contact, ContactGroup, ContactInformation, JmapFile, EmailAddress, Envelope, ContactAddress, ResultReference, JmapResponseError, EmailProperty, EmailBodyProperty))
             handlers.registerType!T;
 
-        static foreach (F; AliasSeq!(getSession, getSessionJson, wellKnownJmap, operatorAsFilter, filterCondition, addQuotes, uniqBy, mailboxPath, allMailboxPaths,
```
Comment (9il) [inline source/jmap/api.d]: `FilterCondition` can be used directly.

File: source/jmap/types.d [hunk 2]
```diff
@@ -877,217 +873,57 @@ struct Invocation {
...
-    import std.string : format;
-    return format!"%04d-%02d-%02dT%02d:%02d:%02dZ"(
-        dt.date.year,
-        dt.date.month,
-        dt.date.day,
-        dt.timeOfDay.hour,
-        dt.timeOfDay.minute,
-        dt.timeOfDay.second,
-    );
+struct FilterOperators {
```
Comment (skoppe) [inline source/jmap/types.d]: Shouldn't this be called `FilterOperator` ?

File: source/jmap/types.d [hunk 3]
```diff
@@ -1124,21 +960,13 @@ class FilterCondition : Filter {
     }
 }
 
-Filter operatorAsFilter(FilterOperator filterOperator) {
```
Comment (skoppe) [inline source/jmap/types.d:1127]: this is still registered at https://github.com/symmetryinvestments/imap-d/pull/109/files#diff-534e4ee761393f362ae7ae3ca4c389bfd507a159a52529ae4bacc2032748f74dR15

File: source/jmap/api.d [hunk 4]
```diff
@@ -9,10 +9,10 @@ version (SIL) {
...
+        static foreach (F; AliasSeq!(getSession, getSessionJson, wellKnownJmap, operatorAsFilter, addQuotes, uniqBy, mailboxPath, allMailboxPaths,
```
Comment (skoppe) [inline source/jmap/api.d]: you removed `filterCondition` but it might be used by @Laeeth

**Review by skoppe — CHANGES_REQUESTED** [review]

File: source/jmap/types.d (hunk 2 above)
Comment (9il) [reply source/jmap/types.d]: Thanks, that is typo

File: source/jmap/api.d (hunk 4 above)
Comment (9il) [reply source/jmap/api.d]: Added deprecation msg instead of removing it.

File: source/jmap/types.d (hunk 3 above)
Comment (9il) [reply source/jmap/types.d:1127]: Thanks, removed from register log.

File: source/jmap/api.d (hunk 1 above)
Comment (skoppe) [reply source/jmap/api.d]: I know, but some code might use `operatorAsFilter` and it will break.

**Review by skoppe — CHANGES_REQUESTED** [review]

**Comment by 9il** [comment]
`operatorAsFilter` has been added back. `FilterHolder` -> `Filter` and I have added constructors for `Filter`.

**Comment by skoppe** [comment]
lovely

**Pull request merged**